lms/views/api/blackboard/sync.py

```diff
@@ -1,17 +1,72 @@
 from pyramid.view import view_config
 
+from lms.models import Grouping
 from lms.security import Permissions
 
 
 class Sync:
     def __init__(self, request):
         self._request = request
+        self._assignment_service = self._request.find_service(name="assignment")
+        self._grouping_service = self._request.find_service(name="grouping")
+        self._course_service = self._request.find_service(name="course")
+        self._blackboard_api = self._request.find_service(name="blackboard_api_client")
+
+        self._tool_consumer_instance_guid = self._request.json["lms"][
+            "tool_consumer_instance_guid"
+        ]
 
     @view_config(
         route_name="blackboard_api.sync",
         request_method="POST",
         renderer="json",
         permission=Permissions.API,
-    )  # pylint: disable=no-self-use
+    )
     def sync(self):
-        return []
+        groups = self._get_blackboard_groups()
+
+        self._sync_to_h(groups)
+
+        authority = self._request.registry.settings["h_authority"]
+        return [group.groupid(authority) for group in groups]
+
+    def _get_blackboard_groups(self):
+        group_set_id = self._group_set()
+        course_id = self._request.json["course"]["context_id"]
+        groups = self._blackboard_api.group_set_groups(course_id, group_set_id)
+
+        return self._to_groups_groupings(groups)
+
+    def _group_set(self):
+        return self._assignment_service.get(
+            self._tool_consumer_instance_guid,
+            self._request.json["course"]["resource_link_id"],
```

Comment:
I think we might want to do all the extracting params from the request at the top of the view method itself. I think it's just nicer to have all the param-reading in one place so you can see which params are actually needed, and this can be near where the schema is defined (once there is a schema):

```python
class Sync:
    ...

    @view_config(..., schema=BlackboardSyncAPISchema)
    def sync(self):
        lms = request.parsed_params["lms"]
        tool_consumer_instance_guid = lms["tool_consumer_instance_guid"]

        course = request.parsed_params["course"]
        context_id = course["context_id"]

        assignment = request.parsed_params["assignment"]
        resource_link_id = assignment["resource_link_id"]

        ...
```

Reply:
By the way, separate PR but I just noticed that the `resource_link_id` is not really part of the course. It's the _assignment_ ID. I think it should be `request.params["assignment"]["resource_link_id"]`. Here's a PR to fix it: <https://github.com/hypothesis/lms/pull/3438> (_this_ PR will need updating after merging that).